Add definitions of tiers-of-support for Wasmtime (#4479)
bytecodealliance/wasmtime: docs/stability-tiers.md

additionally tailored for Wasmtime. Wasmtime's tiered support additionally
applies not only to platforms/targets themselves but additionally features
implemented within Wasmtime itself.

The purpose of this document is to provide a means by which to evaluate the
inclusion of new features and support for existing features within Wasmtime.
This should not be used to "lawyer" a change into Wasmtime on a precise
technical detail or similar since this document is itself not 100% precise and
will change over time.

## Current Tier Status

For explanations of what each tier means see below.

#### Tier 1

* Target - `x86_64-apple-darwin`
* Target - `x86_64-pc-windows-msvc`
* Target - `x86_64-unknown-linux-gnu`
* WASI Proposal - `wasi_snapshot_preview1`
* WASI Proposal - `wasi_unstable`
* WebAssembly Proposal - `bulk-memory`
* WebAssembly Proposal - `reference-types`
* WebAssembly Proposal - `simd`

#### Tier 2

* Target - `aarch64-unknown-linux-gnu`
* Target - `s390x-unknown-linux-gnu`
* Target - `x86_64-pc-windows-gnu`
* WebAssembly Proposal - `memory64`
* WebAssembly Proposal - `multi-memory`

#### Tier 3

* Target - `aarch64-apple-darwin`

> * WASI Proposal - `wasi-nn`

Reviewer: I think wasi-nn could arguably be in tier 2.

Author: Given the current list of requirements sketched out I believe that `wasi-nn` is still largely missing tests. There's one example that is compiled and run but I think that we'd want more of a unit-test-like-suite otherwise.

At least personally this sort of tracks with my gut feeling about `wasi-nn` and `wasi-crypto` where they're present but don't feel all-that-well integrated/first-class. WASI itself isn't really all that first-class by requiring an extra `wasmtime-wasi` crate which hasn't had much time put into its API design for stability too, though. I don't know how best to think about WASI proposals myself.

  Adding a new backend to Cranelift for example is a significant undertaking
  which may not be able to be done in a single PR. Partial implementations of a
  feature are acceptable so long as there's a clear path forward and schedule
  for completing the feature.

* New components in Wasmtime must have a clearly identified owner who is willing
  to be "on the hook" for review, updates to the internals of Wasmtime, etc. For
  example a new backend in Cranelift would need to have a maintainer who is
  willing to respond to changes in Cranelift's interfaces and the needs of
  Wasmtime.

This baseline level of support notably does not require any degree of testing,
fuzzing, or verification. As a result components classified as Tier 3 are
generally not production-ready as they have not been battle-tested much.

Features classified as Tier 3 may be turned off or removed from Wasmtime as
well. If a Tier 3 feature is preventing development of other features then the
owner will be notified. If no response is heard from within a week then the
feature will be turned off. If no further response happens for a month then

> the feature may be removed from Wasmtime.

Reviewer 2: The distinction between "turned off" and "removed" isn't clear in this paragraph. I think you mean that, first, failing CI/fuzz tests will be disabled, and then if the issue still isn't addressed we'll merge a commit removing the implementation entirely. (Although of course that can be reverted later if somebody commits to maintaining it again.) Do I have that right?

Author: Indeed! I'll try to clarify here.

#### Tier 2 - Almost Production Ready

This tier is meant to encompass features and components of Wasmtime which are
well-maintained, tested well, but don't necessarily meet the stringent criteria
for Tier 1. Features in this category may already be "production ready" and safe
to use.

Tier 2 features include:

* Tests are run in CI for the Wasmtime project for this feature and everything
  passes. For example a Tier 2 platform runs in CI directly or via emulation.
  Features are otherwise fully tested on CI.

> * Complete implementations for any feature that's part of Tier 1. For example
>   all Tier 2 targets must implement all of the Tier 1 WebAssembly proposals.

Reviewer 2: For a little more clarity, is this edit accurate?

```suggestion
* Complete implementations for anything that's part of Tier 1. For example
  all Tier 2 targets must implement all of the Tier 1 WebAssembly proposals,
  and all Tier 2 features must be implemented on all Tier 1 targets.
```

that if a component breaks for one reason or another due to an unrelated change
that the maintainer either contributes to the PR-in-progress or otherwise has a
schedule for the implementation of the feature.

#### Tier 1 - Production Ready

This tier is intended to be the highest level of support in Wasmtime for any
particular feature, indicating that it is suitable for production environments.
This conveys a high level of confidence in the Wasmtime project about the
specified features.

Tier 1 features include:

* Continuous fuzzing is required for WebAssembly proposals. This means that any
  WebAssembly proposal must have support in the `wasm-smith` crate and existing
  fuzz targets must be running and exercising the new code paths. Where possible
  differential fuzzing should also be implemented to compare results with other
  implementations.

* Continuous fuzzing is required for the architecture of supported targets. For
  example currently there are three x86\_64 targets that are considered Tier 1
  but only `x86_64-unknown-linux-gnu` is fuzzed.

* CVEs and security releases will be performed as necessary for any bugs found

>   in features.

Reviewer 2: I assume security is part of the requirement for tier 1 targets, not just features?

```suggestion
  in features and targets.
```

> * Most developers of Wasmtime are expected to be able to handle up-to-major
>   changes affecting this component. Specialized maintainers may still be
>   necessary for major refactorings or major new features, but otherwise it
>   should be reasonable to expect most Wasmtime developers to be able to maintain
>   Tier 1 features.

Reviewer 2: I think this rewrite means the same thing but it confuses me less:

```suggestion
* Major changes affecting this component may require help from maintainers with
  specialized expertise, but otherwise it should be reasonable to expect most
  Wasmtime developers to be able to maintain Tier 1 features.
```
